stashcp/__init__.py

```diff
@@ -559,100 +563,184 @@ def get_ips(name):
                 response.close()
             except urllib2.URLError, e:
                 logging.debug("URL error: %s", str(e))
             except Exception, e:
                 logging.debug("Error: %s", str(e))
-            i+=1
+        i+=1
+
+    order_str = ''
+    if len(responselines) > 0:
+        order_str = responselines[0]
         
     if order_str == '':
+        if len(caches_list) == 0:
+            logging.error("unable to get list of caches")
+            return None
         # Unable to find a geo_ip server to use, return random choice from caches!
-        minsite = random.choice(caches_list)['name']
-        random.shuffle(caches_list)
-        nearest_cache_list = [cache['name'] for cache in caches_list]
+        nearest_cache_list = caches_list
+        random.shuffle(nearest_cache_list)
+        minsite = nearest_cache_list[0]
         logging.warning("Unable to use Geoip to find closest cache!  Returning random cache %s", minsite)
-        logging.debug("Ordered list of nearest caches: %s", str(nearest_cache_list))
+        logging.debug("Randomized list of nearest caches: %s", str(nearest_cache_list))
         return minsite
     else:
         # The order string should be something like:
         # 3,1,2
         ordered_list = order_str.strip().split(",")
         logging.debug("Got order %s", str(ordered_list))
-        minsite = caches_list[int(ordered_list[0])-1]['name']
+
+        if len(caches_list) == 0:
+            # Used the stashservers.dat api
+
+            if len(responselines) < 8:
+                logging.error("stashservers response too short, less than 8 lines")
+                return None
+            hashname = responselines[4][-5:]
+            if hashname != "-sha1":
+                logging.error("stashservers response does not have sha1 hash: %s", hashname)
+                return None
+            hashedtext = '\n'.join(responselines[1:5]) + '\n'
+            hash = hashlib.sha1(hashedtext).hexdigest()
+            if responselines[6] != hash:
+                logging.debug("stashservers hash %s does not match expected hash %s", responselines[6], hash)
+                logging.debug("hashed text:\n%s", hashedtext)
+                logging.error("stashservers response hash does not match expected hash")
+                return None
+
+            if not os.path.exists("/usr/bin/openssl"):
+                logging.debug("openssl not installed, skipping signature check")
+            else:
+                sig = '\n'.join(responselines[7:])
+
+                # Look for the OSG cvmfs public key to verify signature
+                prefix = os.environ.get("OSG_LOCATION", "/")
+                osgpub = 'opensciencegrid.org.pub'
+                pubkey_files = ['/etc/cvmfs/keys/opensciencegrid.org/' + osgpub,
+                                os.path.join(prefix, "etc/stashcache", osgpub),
+                                os.path.join(prefix, "usr/share/stashcache", osgpub)]
```

Comment:
Should we add to look wherever this file is located?  As in, it could be beside this script?

Reply:
You're right, I think I was thinking of the `resource_filename` function, which is below.